########################################################################
### Functions for the retrieval of individual data sets from ExpHub
########################################################################

#' Obtain the HPCA data
#'
#' Download and cache the normalized expression values of the data stored in
#' the Human Primary Cell Atlas. The data will be downloaded from ExperimentHub,
#' returning a \linkS4class{SummarizedExperiment} object for further use.
#'
#' @param ensembl Logical scalar indicating whether to convert row names to Ensembl IDs.
#' Genes without a mapping to a non-duplicated Ensembl ID are discarded.
#' @param cell.ont String specifying whether Cell Ontology terms should be included in the \code{\link{colData}}.
#' If \code{"nonna"}, all samples without a valid term are discarded;
#' if \code{"all"}, all samples are returned with (possibly \code{NA}) terms;
#' if \code{"none"}, terms are not added.
#'
#' @details
#' This function provides normalized expression values for 713 microarray samples from
#' the Human Primary Cell Atlas (HPCA) (Mabbott et al., 2013).
#' These 713 samples were processed and normalized as described in Aran, Looney and 
#' Liu et al. (2019).
#' 
#' Each sample has been assigned to one of 37 main cell types (\code{"label.main"}) 
#' and 157 subtypes (\code{"label.fine"}).
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"}, 
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#'
#' @references
#' Mabbott NA et al. (2013).
#' An expression atlas of human primary cells: inference of gene function from coexpression networks.
#' \emph{BMC Genomics} 14, Article 632. 
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- HumanPrimaryCellAtlasData()
#'
#' @export
#' @importFrom SummarizedExperiment rowData
HumanPrimaryCellAtlasData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    version <- "1.0.0"
    se <- .create_se("hpca", version,
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Hs", ensembl)
    se <- .add_ontology(se, "hpca", match.arg(cell.ont))

    se
}

#' Obtain human bulk RNA-seq data from Blueprint and ENCODE
#'
#' Download and cache the normalized expression values of 259 RNA-seq samples of
#' pure stroma and immune cells as generated and supplied by Blueprint and ENCODE.
#'
#' @inheritParams HumanPrimaryCellAtlasData
#' @param rm.NA String specifying how missing values should be handled.
#' \code{"rows"} will remove genes with at least one missing value, 
#' \code{"cols"} will remove samples with at least one missing value,
#' \code{"both"} will remove any gene or sample with at least one missing value,
#' and \code{"none"} will not perform any removal.
#'
#' @details
#' This function provides normalized expression values for 259 bulk RNA-seq samples
#' generated by Blueprint and ENCODE from pure populations of stroma and immune 
#' cells (Martens and Stunnenberg, 2013; The ENCODE Consortium, 2012).
#' The samples were processed and normalized as described in Aran, Looney and
#' Liu et al. (2019), i.e., the raw RNA-seq counts were downloaded from 
#' Blueprint and ENCODE in 2016 and normalized via edgeR (TPMs).
#' 
#' Blueprint Epigenomics contains 144 RNA-seq pure immune samples annotated to 28 cell types.
#' ENCODE contains 115 RNA-seq pure stroma and immune samples annotated to 17 cell types.
#' All together, this reference contains 259 samples with 43 cell types (\code{"label.fine"}),
#' manually aggregated into 24 broad classes (\code{"label.main"}).
#' The fine labels have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#'
#' @references
#' The ENCODE Project Consortium (2012).
#' An integrated encyclopedia of DNA elements in the human genome.
#' \emph{Nature} 489, pages 57–74.
#' 
#' Martens JHA and Stunnenberg HG (2013). 
#' BLUEPRINT: mapping human blood cell epigenomes.
#' \emph{Haematologica} 98, 1487–1489.
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- BlueprintEncodeData(rm.NA = "rows")
#'
#' @export
BlueprintEncodeData <- function(rm.NA = c("rows","cols","both","none"), 
    ensembl=FALSE, cell.ont=c("all", "nonna", "none"))
{
    rm.NA <- match.arg(rm.NA)
    se <- .create_se("blueprint_encode", 
        version = list(logcounts="1.0.0", coldata="1.2.0"),
        assays="logcounts", rm.NA = rm.NA,
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Hs", ensembl)
    se <- .add_ontology(se, "blueprint_encode", match.arg(cell.ont))

    se
}

#' Obtain mouse bulk expression data from the Immunologic Genome Project
#'
#' Download and cache the normalized expression values of 830 microarray samples of
#' pure mouse immune cells, generated by the Immunologic Genome Project (ImmGen).
#'
#' @inheritParams HumanPrimaryCellAtlasData
#'
#' @details
#' This function provides normalized expression values of 830 microarray samples
#' generated by ImmGen from pure populations of murine immune cells (<http://www.immgen.org/>).
#' The samples were processed and normalized as described in Aran, Looney and
#' Liu et al. (2019), i.e., CEL files from the Gene Expression Omnibus (GEO; GSE15907 and GSE37448), 
#' were downloaded, processed, and normalized using the robust multi-array average
#' (RMA) procedure on probe-level data.
#' 
#' This dataset consists of 20 broad cell types (\code{"label.main"}) and 253
#' finely resolved cell subtypes (\code{"label.fine"}).
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#' 
#' @references
#' Heng TS, Painter MW, Immunological Genome Project Consortium (2008).
#' The Immunological Genome Project: networks of gene expression in immune cells.
#' \emph{Nat. Immunol.} 9, 1091-1094. 
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- ImmGenData()
#' 
#' @export
ImmGenData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    se <- .create_se("immgen", version = "1.0.0",
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Mm", ensembl)
    se <- .add_ontology(se, "immgen", match.arg(cell.ont))

    se
}

#' Obtain mouse bulk expression data of sorted cell populations (RNA-seq)
#'
#' Download and cache the normalized expression values of 358 bulk RNA-seq samples
#' of sorted cell populations that can be found at GEO.
#'
#' @inheritParams HumanPrimaryCellAtlasData
#'
#' @details This dataset was contributed by the Benayoun Lab that identified, 
#' downloaded and processed data sets on GEO that corresponded to sorted cell
#' types (Benayoun et al., 2019).
#' 
#' The dataset contains 358 mouse RNA-seq samples annotated to 18 main cell types (\code{"label.main"}):
#' \itemize{
#'     \item Adipocytes
#'     \item Astrocytes
#'     \item B cells 
#'     \item Cardiomyocytes
#'     \item Dendritic cells
#'     \item Endothelial cells
#'     \item Epithelial cells
#'     \item Erythrocytes
#'     \item Fibroblasts
#'     \item Granulocytes
#'     \item Hepatocytes
#'     \item Macrophages
#'     \item Microglia
#'     \item Monocytes
#'     \item Neurons
#'     \item NK cells
#'     \item Oligodendrocytes
#'     \item T cells
#' } 
#' These are split further into 28 subtypes (\code{"label.fine"}).
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#' 
#' @references
#' Benayoun B et al. (2019).
#' Remodeling of epigenome and transcriptome landscapes with aging in mice reveals widespread induction of inflammatory responses.
#' \emph{Genome Res.} 29, 697-709.
#' 
#' Code at \url{https://github.com/BenayounLaboratory/Mouse_Aging_Epigenomics_2018/tree/master/FigureS7_CIBERSORT/RNAseq_datasets_for_Deconvolution/2017-01-18}
#' 
#' @examples
#' ref.se <- MouseRNAseqData()
#' 
#' @export
MouseRNAseqData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    version <- "1.0.0"
    se <- .create_se("mouse.rnaseq", version, 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Mm", ensembl)
    se <- .add_ontology(se, "mouse_rnaseq", match.arg(cell.ont))

    se
}

#' Obtain human bulk RNA-seq data from DICE
#'
#' Download and cache the normalized expression values of 1561 bulk RNA-seq samples
#' of sorted cell populations from the Database of Immune Cell Expression (DICE).
#'
#' @inheritParams HumanPrimaryCellAtlasData
#'
#' @details 
#' This function provides normalized expression values of 1561 bulk RNA-seq samples
#' generated by DICE from pure populations of human immune cells.
#'
#' TPM normalized values for each cell type were downloaded from \url{https://dice-database.org/downloads}. 
#' Genes with no reads across samples were removed, and values were log2 normalized after a pseudocount of 1 was added.
#' 
#' The dataset contains 1561 human RNA-seq samples annotated to 5 main cell types (\code{"label.main"}):
#' \itemize{
#'     \item B cells
#'     \item Monocytes
#'     \item NK cells
#'     \item T cells, CD8+
#'     \item T cells, CD4+
#' }
#'
#' Samples were additionally annotated to 15 fine cell types (\code{"label.fine"}):
#' \itemize{
#'     \item B cells, naive
#'     \item Monocytes, CD14+
#'     \item Monocytes, CD16+
#'     \item NK cells
#'     \item T cells, memory TREG
#'     \item T cells, CD4+, naive
#'     \item T cells, CD4+, naive, stimulated
#'     \item T cells, CD4+, naive Treg
#'     \item T cells, CD4+, Th1
#'     \item T cells, CD4+, Th1_17
#'     \item T cells, CD4+, Th2
#'     \item T cells, CD8+, naïve
#'     \item T cells, CD8+, naïve, stimulated
#'     \item T cells, CD4+, TFH
#'     \item T cells, CD4+, Th17
#' }
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Schmiedel B et al. (2018).
#' Impact of Genetic Polymorphisms on Human Immune Cell Gene Expression.
#' \emph{Cell} 175, 1701-1715.
#' 
#' @examples
#' ref.se <- DatabaseImmuneCellExpressionData()
#' 
#' @export
DatabaseImmuneCellExpressionData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    version <- "1.0.0"
    se <- .create_se("dice", version, 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Hs", ensembl)
    se <- .add_ontology(se, "dice", match.arg(cell.ont))

    se
}

#' Obtain bulk microarray expression for sorted hematopoietic cells 
#'
#' Download and cache the normalized expression values of 211 bulk human microarray samples
#' of sorted hematopoietic cell populations that can be found in 
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24759}{GSE24759}.
#'
#' @inheritParams HumanPrimaryCellAtlasData
#'
#' @details 
#' The dataset contains 211 human microarray samples annotated to 16 main cell types (\code{"label.main"}):
#' \itemize{
#'     \item Basophils
#'     \item B cells
#'     \item CMPs
#'     \item Dendritic cells
#'     \item Eosinophils
#'     \item Erythroid cells
#'     \item GMPS
#'     \item Granulocytes
#'     \item HSCs
#'     \item Megakaryocytes
#'     \item MEPs
#'     \item Monocytes
#'     \item NK cells
#'     \item NK T cells
#'     \item CD8+ T cells
#'     \item CD4+ T cells
#' }
#'
#' Samples were additionally annotated to 38 fine cell types (\code{"label.fine"}):
#' \itemize{
#'     \item Basophils
#'     \item Naive B cells
#'     \item Mature B cells class able to switch
#'     \item Mature B cells
#'     \item Mature B cells class switched
#'     \item Common myeloid progenitors
#'     \item Plasmacytoid Dendritic Cells
#'     \item Myeloid Dendritic Cells
#'     \item Eosinophils
#'     \item Erythroid_CD34+ CD71+ GlyA-
#'     \item Erythroid_CD34- CD71+ GlyA-
#'     \item Erythroid_CD34- CD71+ GlyA+
#'     \item Erythroid_CD34- CD71lo GlyA+
#'     \item Erythroid_CD34- CD71- GlyA+
#'     \item Granulocyte/monocyte progenitors
#'     \item Colony Forming Unit-Granulocytes
#'     \item Granulocyte (Neutrophilic Metamyelocytes)
#'     \item Granulocyte (Neutrophils)
#'     \item Hematopoietic stem cells_CD133+ CD34dim
#'     \item Hematopoietic stem cell_CD38- CD34+
#'     \item Colony Forming Unit-Megakaryocytic
#'     \item Megakaryocytes
#'     \item Megakaryocyte/erythroid progenitors
#'     \item Colony Forming Unit-Monocytes
#'     \item Monocytes
#'     \item Mature NK cells_CD56- CD16+ CD3-
#'     \item Mature NK cells_CD56+ CD16+ CD3-
#'     \item Mature NK cells_CD56- CD16- CD3-
#'     \item NK T cells
#'     \item Early B cells
#'     \item Pro B cells
#'     \item CD8+ Effector Memory RA
#'     \item Naive CD8+ T cells
#'     \item CD8+ Effector Memory
#'     \item CD8+ Central Memory
#'     \item Naive CD4+ T cells
#'     \item CD4+ Effector Memory
#'     \item CD4+ Central Memory
#' }
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Novershtern N et al. (2011).
#' Densely interconnected transcriptional circuits control cell states in human hematopoiesis.
#' \emph{Cell} 144, 296-309.
#' 
#' @examples
#' ref.se <- NovershternHematopoieticData()
#' 
#' @export
NovershternHematopoieticData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    se <- .create_se("dmap", 
        version = c(logcounts="1.0.0", coldata="1.2.0"), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Hs", ensembl)
    se <- .add_ontology(se, "novershtern", match.arg(cell.ont))

    se
}

#' Obtain bulk RNA-seq data of sorted human immune cells
#'
#' Download and cache the normalized expression values of 114 bulk RNA-seq samples
#' of sorted immune cell populations that can be found in 
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011}{GSE107011}.
#'
#' @inheritParams HumanPrimaryCellAtlasData
#'
#' @details 
#' The dataset contains 114 human RNA-seq samples annotated to 10 main cell types (\code{"label.main"}):
#' \itemize{
#'     \item CD8+ T cells
#'     \item T cells
#'     \item CD4+ T cells
#'     \item Progenitors
#'     \item B cells
#'     \item Monocytes
#'     \item NK cells
#'     \item Dendritic cells
#'     \item Neutrophils
#'     \item Basophils
#' }
#'
#' Samples were additionally annotated to 29 fine cell types (\code{"label.fine"}):
#' \itemize{
#'     \item Naive CD8 T cells
#'     \item Central memory CD8 T cells
#'     \item Effector memory CD8 T cells
#'     \item Terminal effector CD8 T cells
#'     \item MAIT cells
#'     \item Vd2 gd T cells
#'     \item Non-Vd2 gd T cells
#'     \item Follicular helper T cells
#'     \item T regulatory cells
#'     \item Th1 cells
#'     \item Th1/Th17 cells
#'     \item Th17 cells
#'     \item Th2 cells
#'     \item Naive CD4 T cells
#'     \item Terminal effector CD4 T cells
#'     \item Progenitor cells
#'     \item Naive B cells
#'     \item Non-switched memory B cells
#'     \item Exhausted B cells
#'     \item Switched memory B cells
#'     \item Plasmablasts
#'     \item Classical monocytes
#'     \item Intermediate monocytes
#'     \item Non classical monocytes
#'     \item Natural killer cells
#'     \item Plasmacytoid dendritic cells
#'     \item Myeloid dendritic cells
#'     \item Low-density neutrophils
#'     \item Low-density basophils
#' }
#' The subtypes have also been mapped to the Cell Ontology (\code{"label.ont"},
#' if \code{cell.ont} is not \code{"none"}), which can be used for further programmatic
#' queries.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Monaco G et al. (2019).
#' RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types
#' \emph{Cell Rep.} 26, 1627-1640.
#' 
#' @examples
#' ref.se <- MonacoImmuneData()
#' 
#' @export
MonacoImmuneData <- function(ensembl=FALSE, cell.ont=c("all", "nonna", "none")) {
    version <- "1.0.0"
    se <- .create_se("monaco_immune", version, 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)

    se <- .convert_to_ensembl(se, "Hs", ensembl)
    se <- .add_ontology(se, "monaco", match.arg(cell.ont))

    se
}

#####################################################################
### Helper function
#####################################################################

#' @importFrom ExperimentHub ExperimentHub
#' @importFrom SummarizedExperiment SummarizedExperiment
#' @importFrom SummarizedExperiment rowData
#' @importFrom S4Vectors DataFrame
.create_se <- function(dataset, version, hub = ExperimentHub(), assays="logcounts",
    rm.NA = c("rows","cols","both","none"), has.rowdata=FALSE, has.coldata=TRUE) 
{
    rm.NA <- match.arg(rm.NA)
    host <- file.path("SingleR", dataset)

    ## extract normalized values --------
    all.assays <- list()
    for (a in assays) {
        ver <- .fetch_version(version, a)
        nrmcnts <- hub[hub$rdatapath==file.path(host, ver, paste0(a, ".rds"))][[1]]
        all.assays[[a]] <- .rm_NAs(nrmcnts, rm.NA)
    }
    
    ## get metadata ----------------------
    args <- list()
    if (has.coldata) {
        ver <- .fetch_version(version, "coldata")
        args$colData <- hub[hub$rdatapath==file.path(host, ver, "coldata.rds")][[1]]
    }
    if (has.rowdata) {
        ver <- .fetch_version(version, "rowdata")
        args$rowData <- hub[hub$rdatapath==file.path(host, ver, "rowdata.rds")][[1]]
    }
    
    ## make the final SE object ----------
    do.call(SummarizedExperiment, c(list(assays=all.assays), args))
}

.fetch_version <- function(version, field) {
    opt <- version[field]
    if (is.na(opt)) {
        version[1]
    } else {
        opt
    }
}

#' @importFrom DelayedMatrixStats rowAnyNAs colAnyNAs
#' @importFrom DelayedArray DelayedArray
.rm_NAs <- function(mat, rm.NA = "rows"){
    # Identify them first before removal, to ensure that 
    # the same rows are columns are removed with 'both'.
    if(rm.NA == "rows" || rm.NA == "both"){
        keep_rows <- !rowAnyNAs(DelayedArray(mat))
    } else {
        keep_rows <- !logical(nrow(mat))
    }

    if (rm.NA=="cols" || rm.NA == "both") {
        keep_cols <- !colAnyNAs(DelayedArray(mat))
    } else {
        keep_cols <- !logical(ncol(mat))
    }

    # Avoid making unnecessary copies if possible.
    if (!all(keep_rows)) {
        mat <- mat[keep_rows,,drop=FALSE]
    }
    if (!all(keep_cols)) {
        mat <- mat[,keep_cols,drop=FALSE]
    }
    mat
}

.convert_to_ensembl <- function(se, species, ensembl) 
# Copied verbatim from scRNAseq... well, whatever.
{
    if (ensembl) {
        if (species=="Mm") {
            tag <- "AH73905"
        } else if (species=="Hs") {
            tag <- "AH73881"
        }

        edb <- AnnotationHub::AnnotationHub()[[tag]]
        ensid <- AnnotationDbi::mapIds(edb, keys=rownames(se), keytype="SYMBOL", column="GENEID")

        keep <- !is.na(ensid) & !duplicated(ensid)
        if (!all(keep)) {
            se <- se[keep,]
        }
        rownames(se) <- ensid[keep]
    }

    se
}

#' @importFrom utils read.delim
.add_ontology <- function(se, fname, mode) {
    if (mode!="none") {
        src <- read.delim(system.file("mapping", paste0(fname, ".tsv"), 
            package="SingleR", mustWork=TRUE), header=FALSE, stringsAsFactors=FALSE)
        m <- match(se$label.fine, src[,1])
        stopifnot(all(!is.na(m))) # sanity check

        matched <- src[m, 2]
        matched[matched==""] <- NA_character_
        se$label.ont <- matched

        if (mode=="nonna") {
            se <- se[,!is.na(se$label.ont)]
        }
    }

    se
}
